memcache/stats.go

```diff
@@ -0,0 +1,474 @@
+	// SlabReassignRescues Items rescued from eviction in page move.
+	SlabReassignRescues uint64
+	// SlabReassignEvictionsNomem Valid items evicted during a page move (due to no free memory in slab).
+	SlabReassignEvictionsNomem uint64
+	// SlabReassignChunkRescues Individual sections of an item rescued during a page move.
+	SlabReassignChunkRescues uint64
+	// SlabReassignInlineReclaim Internal stat counter for when the page mover clears memory from the chunk freelist when it wasn't expecting to.
+	SlabReassignInlineReclaim uint64
+	// SlabReassignBusyItems Items busy during page move, requiring a retry before page can be moved.
+	SlabReassignBusyItems uint64
+	// SlabReassignBusyDeletes Items busy during page move, requiring deletion before page can be moved.
+	SlabReassignBusyDeletes uint64
+	// LogWorkerDropped Logs a worker never wrote due to full buf.
+	LogWorkerDropped uint64
+	// LogWorkerWritten Logs written by a worker, to be picked up.
+	LogWorkerWritten uint64
+	// LogWatcherSkipped Logs not sent to slow watchers.
+	LogWatcherSkipped uint64
+	// LogWatcherSent Logs written to watchers.
+	LogWatcherSent uint64
+}
+
+// StatsServers returns the general statistics of the servers
+// retrieved with the `stats` command
+func (c *Client) StatsServers() (servers map[net.Addr]*ServerStats, err error) {
+	// Each ServerStats has its own "asociated" sync.Mutex.
+	servers = make(map[net.Addr]*ServerStats)
+	muxes := make(map[net.Addr]*sync.Mutex)
+
+	// addrs slice of the all the server adresses from the selector.
+	addrs := make([]net.Addr, 0)
+
+	// For each server addres a ServerStats, sync.mutex  is created.
+	err = c.selector.Each(
+		func(addr net.Addr) error {
+			addrs = append(addrs, addr)
+			servers[addr] = new(ServerStats)
+			muxes[addr] = new(sync.Mutex)
+			return nil
+		},
+	)
+	if err != nil {
+		return
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(len(addrs))
+	for _, addr := range addrs {
+		go func(addr net.Addr) {
+			mux := muxes[addr]
+			server := servers[addr]
+			c.statsFromAddr(server, mux, addr)
+			wg.Done()
+		}(addr)
+	}
+	wg.Wait()
+	return
+}
+
+func (c *Client) statsFromAddr(server *ServerStats, mux *sync.Mutex, addr net.Addr) {
+	var wg sync.WaitGroup
+	err := c.withAddrRw(addr, func(rw *bufio.ReadWriter) error {
+		if _, err := fmt.Fprintf(rw, "stats\r\n"); err != nil {
+			return err
+		}
+		if err := rw.Flush(); err != nil {
+			return err
+		}
+
+		// STAT <name> <value>\r\n
+		for {
+			line, err := rw.ReadBytes('\n')
+			if err != nil {
+				return err
+			}
+			if bytes.Equal(line, resultEnd) {
+				return nil
+			}
+
+			wg.Add(1)
+			go func() {
+				server.parseStatValue(string(line), mux)
+				wg.Done()
+			}()
+		}
+	})
+	wg.Wait()
+	server.ServerErr = err
+	return
+}
+
+func (server *ServerStats) parseStatValue(line string, mux *sync.Mutex) {
+	var err error
+	tkns := strings.Split(line, " ")
+	v := tkns[2][:len(tkns[2])-2]
+	mux.Lock()
+	defer mux.Unlock()
+	switch tkns[1] {
+	case "version":
```

Comment:
This whole switch is pretty repetitive. Can we just struct tags on the struct above and instead use reflect? (ala encoding/json, etc)

Reply:
Thanks, it didn't occur to me to do it with reflection.

I converted the snake case stat name to camel case, and thats the name I use in the structs, and with that, I only have to call the method **FieldByName**